lib/awful/screenshot.lua

```diff
@@ -0,0 +1,884 @@
+--
+-- @author Brian Sobulefsky &lt;brian.sobulefsky@protonmail.com&gt;
+-- @copyright 2021 Brian Sobulefsky
+-- @inputmodule awful.screenshot
+---------------------------------------------------------------------------
+
+-- Grab environment we need
+local capi = { root = root,
+               screen = screen,
+               client = client,
+               mousegrabber = mousegrabber }
+local aw_screen = require("awful.screen")
+local gears = require("gears")
+local beautiful = require("beautiful")
+local wibox = require("wibox")
+local cairo = require("lgi").cairo
+local naughty = require("naughty")
+
+-- The module to be returned
+local module = { mt = {} }
+-- The screenshow object created by a call to the screenshow module
+local screenshot = {}
+-- The various methods of taking a screenshow (root window, client, etc).
+local screenshot_methods = {}
+
+-- Configuration data
+local ss_dir = nil
+local ss_prfx = nil
+local frame_color = gears.color("#000000")
+local initialized = nil
+
+
+-- Internal function to get the default screenshot directory
+-- 
+-- Can be expanded to read X environment variables or configuration files.
+local function get_default_dir()
+  -- This can be expanded
+  local d = os.getenv("HOME")
+  
+  if d then
+    d = string.gsub(d, '/*$', '/') .. 'Images/'
+    if os.execute("bash -c \"if [ -d \\\"" .. d .. "\\\" -a -w \\\"" .. d ..
+                    "\\\" ] ; then exit 0 ; else exit 1 ; fi ;\"") then
+      return d
+    end
+  end
+
+  return nil
+end
+
+-- Internal function to get the default filename prefix
+local function get_default_prefix()
+  return "Screenshot-"
+end
+
+-- Internal function to check a directory string for existence and writability.
+-- This only checks if the requested directory exists and is writeable, not if
+-- such a directory is legal (i.e. it behaves as 'mkdir' and not 'mkdir -p').
+-- Adding 'mkdir -p' functionality can be considered in the future.
+local function check_directory(directory)
+
+  if directory and type(directory) == "string"  then
+
+    -- Fully qualify a "~/" path or a relative path to $HOME. One might argue
+    -- that the relative path should fully qualify to $PWD, but for a window
+    -- manager this should be the same thing to the extent that anything else
+    -- is arguably unexpected behavior.
+    if string.find(directory, "^~/") then
+      directory = string.gsub(directory, "^~/",
+                              string.gsub(os.getenv("HOME"), "/*$", "/"))
+    elseif string.find(directory, "^[^/]") then
+      directory = string.gsub(os.getenv("HOME"), "/*$", "/") .. directory
+    end
+
+    -- Assure that we return exactly one trailing slash
+    directory = string.gsub(directory, '/*$', '/')
+
+    -- If we use single quotes, we only need to deal with single quotes - (I
+    -- promise that's meaningful if you think about it from a bash perspective)
+    local dir = string.gsub(directory, "'", "'\\'\\\\\\'\\''")
```

Comment:
what the difference between dir and directory??

Reply:
So that one is actually for a reason but I can do a better name. That is the escaped string to be passed through bash.

Reply:
that should be clear from var name - not from your comment

it makes no sense giving basically the same name to 2 variables which have different logical meaning